src/vm/class.cpp

```diff
@@ -1731,6 +1731,17 @@ EEClass::CheckForHFA()
         if ((strcmp(className, "Vector256`1") == 0) || (strcmp(className, "Vector128`1") == 0) ||
             (strcmp(className, "Vector64`1") == 0))
         {
+            if (IsCompilationProcess())
```

Comment:
Why is not the existing check in methodtablebuilder.cpp sufficient?

Reply:
This method is specific to platforms with HFA. If there is a problem with nested structs, it apply for all platforms; not just to platforms with HFAs.

Reply:
> Why is not the existing check in methodtablebuilder.cpp sufficient?

I am not sure.  Here are my guesses:
+ crossgen of the Vector128<T> et al. Debug Views
+ crossgen of `System.Runtime.Intrinsics.X86` is not disabled on Arm64, so it gets here...

> This method is specific to platforms with HFA

Yes this is definitely not a complete cross platform solution.

Reply:
This is the back trace when running crossgen on S.P.C.

```
#0  EEClass::CheckForHFA (this=0xffffafcd6290, pByValueClassCache=0x0) at /home/vmjenkins/workspace/Dotnet/build_and_test/src/vm/cla
ss.cpp:1787
#1  0x00000000007252d4 in MethodTableBuilder::BuildMethodTableThrowing (this=0xffffffffc630, pAllocator=0xd64380 <g_pSystemDomainMem
ory+3792>, pLoaderModule=0xffffb71b0020, pModule=0xffffb71b0020, cl=33555485, pBuildingInterfaceList=0x0, pLayoutRawFieldInfos=0x0,
pParentMethodTable=0xffffb71b7ca0, bmtGenericsInfo=0xffffffffc7f0, parentInst=..., cBuildingInterfaceList=0) at /home/vmjenkins/work
space/Dotnet/build_and_test/src/vm/methodtablebuilder.cpp:1884
#2  0x000000000073cc38 in ClassLoader::CreateTypeHandleForTypeDefThrowing (pModule=0xffffb71b0020, cl=33555485, inst=..., pamTracker
=0xffffffffcdc0) at /home/vmjenkins/workspace/Dotnet/build_and_test/src/vm/methodtablebuilder.cpp:12178
#3  0x00000000008323e0 in ClassLoader::CreateTypeHandleForTypeKey (pKey=0xffffffffd4a8, pamTracker=0xffffffffcdc0) at /home/vmjenkin
s/workspace/Dotnet/build_and_test/src/vm/clsload.cpp:3589
#4  0x00000000008320a4 in ClassLoader::DoIncrementalLoad (pTypeKey=0xffffffffd4a8, typeHnd=..., currentLevel=CLASS_LOAD_BEGIN) at /h
ome/vmjenkins/workspace/Dotnet/build_and_test/src/vm/clsload.cpp:3517
#5  0x0000000000833ce4 in ClassLoader::LoadTypeHandleForTypeKey_Body (this=0xdc93f0, pTypeKey=0xffffffffd4a8, typeHnd=..., targetLev
el=CLASS_LOAD_EXACTPARENTS) at /home/vmjenkins/workspace/Dotnet/build_and_test/src/vm/clsload.cpp:4325
#6  0x000000000082e1a4 in ClassLoader::LoadTypeHandleForTypeKey (this=0xdc93f0, pTypeKey=0xffffffffd4a8, typeHnd=..., targetLevel=CL
ASS_LOADED, pInstContext=0x0) at /home/vmjenkins/workspace/Dotnet/build_and_test/src/vm/clsload.cpp:4044
#7  0x000000000082ffac in ClassLoader::LoadTypeDefThrowing (pModule=0xffffb71b0020, typeDef=33555485, fNotFoundAction=ClassLoader::T
hrowIfNotFound, fUninstantiated=ClassLoader::PermitUninstDefOrRef, tokenNotToLoad=0, level=CLASS_LOADED, pTargetInstantiation=0x0) a
t /home/vmjenkins/workspace/Dotnet/build_and_test/src/vm/clsload.cpp:2887
#8  0x000000000083126c in ClassLoader::LoadTypeDefOrRefThrowing (pModule=0xffffb71b0020, typeDefOrRef=33555485, fNotFoundAction=Clas
sLoader::ThrowIfNotFound, fUninstantiated=ClassLoader::PermitUninstDefOrRef, tokenNotToLoad=0, level=CLASS_LOADED) at /home/vmjenkin
s/workspace/Dotnet/build_and_test/src/vm/clsload.cpp:3065
#9  0x00000000007fa6bc in LoadTypeDefOrRefHelper (image=0xe0d740, pModule=0xffffb71b0020, tk=33555485) at /home/vmjenkins/workspace/
Dotnet/build_and_test/src/vm/ceeload.cpp:6883
#10 0x00000000007f9ad0 in Module::ExpandAll (this=0xffffb71b0020, image=0xe0d740) at /home/vmjenkins/workspace/Dotnet/build_and_test
/src/vm/ceeload.cpp:7747
#11 0x00000000006626d4 in CEEPreloader::Preload (this=0xe0d660, profileData=0xe0d250) at /home/vmjenkins/workspace/Dotnet/build_and_
test/src/vm/compile.cpp:4638
#12 0x00000000006625b0 in CEECompileInfo::PreloadModule (this=0xd9add0, module=0xffffb71b0020, pData=0xe0bf00, profileData=0xe0d250)
 at /home/vmjenkins/workspace/Dotnet/build_and_test/src/vm/compile.cpp:4572
#13 0x0000000000950dd4 in ZapImage::Preload (this=0xe0be00) at /home/vmjenkins/workspace/Dotnet/build_and_test/src/zap/zapimage.cpp:
1411
#14 0x0000000000949ccc in Zapper::CompileModule (this=0xd9ab60, hModule=0xffffb71b0020, pAssemblyEmit=0xdd1558) at /home/vmjenkins/w
orkspace/Dotnet/build_and_test/src/zap/zapper.cpp:1659
```
It hits twice for each of these.
```
1: namespaceName = (LPCUTF8) 0xffffb4aaf953 "System.Runtime.Intrinsics"
2: className = (LPCUTF8) 0xffffb4a9e729 "Vector64`1"
```
```
1: namespaceName = (LPCUTF8) 0xffffb4aaf953 "System.Runtime.Intrinsics"
2: className = (LPCUTF8) 0xffffb4a9e740 "Vector128`1"
```

Looks like it never gets to code in MethodTableBuilder.cpp.  Before it hits in CheckHFA(), last time in methodtablebuilder is.
```
3: className = (LPCUTF8) 0xffffb4aa7271 "HandleProcessCorruptedStateExceptionsAttribute"
4: nameSpace = (LPCUTF8) 0xffffb4aafd28 "System.Runtime.ExceptionServices"
```

Reply:
I think this check should be in `CheckForSystemTypes` in methodtable


```
                    _ASSERTE_MSG(FALSE, "Unhandled Hardware Intrinsic Type.");
                }

>>>         if (IsCompilationProcess()) throw;

                return;
```

Reply:
If I put a break point in both.  `MethodTableBuilder::CheckForSystemTypes()` is later in the crossgen process.

Seems like it needs to be in `GetHalfBakedClass()` or the construction of a half baked class.  We need to disable crossgen of any methods which pass class by value.

Reply:
> MethodTableBuilder::CheckForSystemTypes() is later in the crossgen process.

That should be fine. Throwing in this method should still cause the type to fail to load.

Reply:
MethodTableBuilder::CheckForSystemTypes()@jkotas Adding the check to `MethodTableBuilder::CheckForSystemTypes()` seems to work.  Do you want to remove the other one in methodtablebuilder.cpp?

Reply:
Yes, it would be nice to not have redundant/dead checks for this.

Reply:
@jkotas Done.  Moved to `MethodTableBuilder::CheckForSystemTypes()`  removed redundant check in `MethodTableBuilder::BuildMethodTableThrowing()`

This should be pulled into 2.1 release branch.

Reply:
> This should be pulled into 2.1 release branch.

It will get there for preview2.